### PR TITLE
Bugfix: order product list by position

### LINKED ISSUE #3058: Category sort = 'position:asc' does not reflect the order set within magento categories
## Current behavior
Currently it does not seem possible to display products within categories using the order specified in the Magento admin. 

I am able to see the position number in the elastic data using Kibana. When using the new magento2-vsbridge-indexer you can see this in the categories attribute against a product. 

When using the mage2vuestorefront to import data, the position number is set on extension_attributes.category_links

In either case, setting sort = 'position:asc' or 'position:desc' makes no difference to the sorting of products.

## Expected behavior
When sort = 'position:asc' products within categories should be ordered as per the order field within magento admin

## Steps to reproduce the issue
Either using postman to hit the api directly or by configuring the storefront to sort products by position:asc products will not sort by the position number. And position:desc returns products in same position as position:asc. 

## Can you handle fixing this bug by yourself?
* [ ] YES
* [x] NO

## Which [Release Cycle]
* [ ] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
* [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
* [x] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

### PR DESCRIPTION
**Pull request by rozzilla**
### Related Issues
closes #3058

### Short Description of the PR
Bugfix: order product list by position, for Magento 1 & 2.

### Pull Request Checklist

- [x] I have updated the Changelog ([V1](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md)) [v2](https://docs-next.vuestorefront.io/contributing/creating-changelog.html) and mentioned all breaking changes in the public API.
- [x] I have documented all new public APIs and made changes to existing docs mentioning the parts I've changed so they're up to date.
- [x] I have tested my Pull Request on production build and (to my knowledge) it works without any issues

### CODE REVIEW

**Comment by rozzilla** [comment]
@Fifciu, @lukaszjedrasik, please let me know if now is everything ok, thanks!

**Review by lukaszjedrasik — APPROVED** [review]

File: core/lib/search/adapter/api/searchAdapter.ts [hunk 1]
```diff
@@ -8,7 +8,8 @@ import { SearchQuery } from 'storefront-query-builder'
 import HttpQuery from '@vue-storefront/core/types/search/HttpQuery'
 import { SearchResponse } from '@vue-storefront/core/types/search/SearchResponse'
 import config from 'config'
-import getApiEndpointUrl from '@vue-storefront/core/helpers/getApiEndpointUrl';
+import getApiEndpointUrl from '@vue-storefront/core/helpers/getApiEndpointUrl'
+import store from '../../../../store'
```
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:12]:
```suggestion
import rootStore from '@vue-storefront/core/store';
```

File: core/lib/search/adapter/api/searchAdapter.ts [hunk 2]
```diff
@@ -75,14 +76,36 @@ export class SearchAdapter {
     url = url + '/' + encodeURIComponent(Request.index) + '/' + encodeURIComponent(Request.type) + '/_search'
     url = url + '?' + queryString.stringify(httpQuery)
 
+    let queryMethod = config.elasticsearch.queryMethod
+    if (Request.type === 'product' && Request.sort) {
+      const sortType = Request.sort.split(':', 2)
+      if (sortType[0] === 'position') {
+        const currentCategoryId = store.getters['category-next/getCurrentCategory']['id']
```
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:83]:
```suggestion
        const currentCategoryId = rootStore.getters['category-next/getCurrentCategory']['id']
```

File: core/lib/search/adapter/api/searchAdapter.ts [hunk 3]
```diff
@@ -75,14 +76,36 @@ export class SearchAdapter {
     url = url + '/' + encodeURIComponent(Request.index) + '/' + encodeURIComponent(Request.type) + '/_search'
     url = url + '?' + queryString.stringify(httpQuery)
 
+    let queryMethod = config.elasticsearch.queryMethod
+    if (Request.type === 'product' && Request.sort) {
+      const sortType = Request.sort.split(':', 2)
```
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:81]:
```suggestion
      const [ sortByField, sortOrder ] = Request.sort.split(':', 2)
```
Remember to update in different places

File: core/lib/search/adapter/api/searchAdapter.ts (hunk 2 above)
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:83]:
```suggestion
        const currentCategoryId = store.getters['category-next/getCurrentCategory'].id
```

File: core/lib/search/adapter/api/searchAdapter.ts [hunk 4]
```diff
@@ -75,14 +76,36 @@ export class SearchAdapter {
...
+      if (sortType[0] === 'position') {
+        const currentCategoryId = store.getters['category-next/getCurrentCategory']['id']
+        if (currentCategoryId) {
+          // It's mandatory to send a POST request, in this case, otherwise, this sorting won't work
```
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:85]: Why? We won't benefit from browser's caching

File: core/lib/search/adapter/api/searchAdapter.ts [hunk 5]
```diff
@@ -75,14 +76,36 @@ export class SearchAdapter {
...
+          queryMethod = 'POST'
+          ElasticsearchQueryBody['sort'] = [{
+            'category.position': {
+              'order': sortType[1] || 'asc',
```
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:89]: Not needed `'` for most of fields

File: config/default.json [hunk 6]
```diff
@@ -612,7 +612,9 @@
     "sortByAttributes": {
       "Latest": "updated_at:desc",
       "Price: Low to high": "final_price",
-      "Price: High to low": "final_price:desc"
+      "Price: High to low": "final_price:desc",
+      "Position: First to last": "position:asc",
```
Comment (Fifciu) [inline config/default.json:616]:
```suggestion
      "Position: First to last": "position",
```
`asc` is rather default one

File: core/lib/search/adapter/api/searchAdapter.ts [hunk 7]
```diff
@@ -75,14 +76,36 @@ export class SearchAdapter {
     url = url + '/' + encodeURIComponent(Request.index) + '/' + encodeURIComponent(Request.type) + '/_search'
     url = url + '?' + queryString.stringify(httpQuery)
 
+    let queryMethod = config.elasticsearch.queryMethod
+    if (Request.type === 'product' && Request.sort) {
```
Comment (Fifciu) [inline core/lib/search/adapter/api/searchAdapter.ts:80]: As `asc` sorting is default one. You should make support for this in these 2 lines (80-81)

**Review by Fifciu — CHANGES_REQUESTED** [review]

File: core/lib/search/adapter/api/searchAdapter.ts (hunk 4 above)
Comment (rozzilla) [reply core/lib/search/adapter/api/searchAdapter.ts:85]: Because otherwise, with a GET request, it doesn't work.

**Comment by rozzilla** [comment]
I've updated the PR with your suggestions @Fifciu, please take a look :)

**Comment by rozzilla** [comment]
Any news abou it, @Fifciu ?

**Review by Fifciu — APPROVED** [review]
For me, it looks well, great job. Thanks a lot for this fabulous feature @rozzilla !

**Comment by rozzilla** [comment]
Just to know, @Fifciu, what's the roadmap for merge into master? I have
another PR merged form many months right now, but it's still in develop...

Il lun 21 giu 2021, 09:53 Filip Jędrasik ***@***.***> ha
scritto:

**Comment by Fifciu** [comment]
Currently, we are working on release 1.12.3, then we will be focused on solving issues related to 1.13 and releasing it

**Comment by rozzilla** [comment]
Ok, thanks for your response

Il lun 21 giu 2021, 16:52 Filip Jędrasik ***@***.***> ha
scritto:

**Comment by antonioglass** [comment]
Tried it but didn't work. Had to change category_category_id to 'category.category_id'. Now it works.

**Comment by antonioglass** [comment]
@rozzilla Hi! Thanks very much for the fix but I faced a small issue. Have you also faced it? See here: https://github.com/vuestorefront/vue-storefront-1/issues/533 Thanks in advance!

**Comment by rozzilla** [comment]
> @rozzilla Hi! Thanks very much for the fix but I faced a small issue. Have you also faced it? See here: [vuestorefront/vue-storefront-1#533](https://github.com/vuestorefront/vue-storefront-1/issues/533) Thanks in advance!

It's quite a while I'm not working on VSF so a lot of things change in the meantime. Feel free to change this part of code ;)

**Comment by antonioglass** [comment]
Fixed! The wrong category id was returned after switching to another category.

Replace this:
`      if (sortByField === 'position') {
        const currentCategoryId = rootStore.getters['category-next/getCurrentCategory'].id`

With this:
`      const categories = Request.searchQuery.getAppliedFilters().find(element => element.attribute === 'category_ids')
      if (sortByField === 'position' && categories && categories.value) {
        const currentCategoryId = categories.value.in[0]`

**Pull request merged**